lib/puppet/application.rb

```diff
@@ -281,6 +281,12 @@ def run_mode( mode_name = nil)
       @run_mode = Puppet::Util::RunMode[ mode_name || Puppet.settings.preferred_run_mode ]
     end
 
+    def environment_mode( mode_name = nil)
+      mode_name = mode_name.to_s.downcase.intern || :local
+      raise ValidationError, "Invalid run mode '#{mode}'" unless [:local, :remote, :not_required].include?(mode_name)
```

Comment:
Should this be `mode_name` instead of `mode`?

Reply:
Found a few issues here. Working with Melissa to fix it up.